# [M3S03] - Ex. 3 — Validação Cruzada (cross_val_score)

**Objetivo:** aplicar validação cruzada com 5 folds no modelo de Regressão Linear e verificar se o resultado é consistente com o teste simples (hold-out 80/20).

**Base de dados:** `dados_AP - [M3S01] - Ex. 1 até [M3S03] - Ex. 4.csv`

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score

## 1. Preparação dos dados

In [2]:
df = pd.read_csv('../../data/input/csv/dados_AP - [M3S01] - Ex. 1 até [M3S03] - Ex. 4.csv')

X = df[['investimento_marketing', 'preco_medicamento', 'mes']]
y = df['demanda']

## 2. Validação cruzada — 5 folds

In [3]:
modelo = LinearRegression()

scores_r2   = cross_val_score(modelo, X, y, cv=5, scoring='r2')
scores_mae  = cross_val_score(modelo, X, y, cv=5, scoring='neg_mean_absolute_error')
scores_rmse = cross_val_score(modelo, X, y, cv=5, scoring='neg_root_mean_squared_error')

print('R² por fold :', np.round(scores_r2, 4))
print()
print(f'R²   — média: {scores_r2.mean():.4f}  | desvio padrão: {scores_r2.std():.4f}')
print(f'MAE  — média: {(-scores_mae).mean():.2f} | desvio padrão: {(-scores_mae).std():.2f}')
print(f'RMSE — média: {(-scores_rmse).mean():.2f} | desvio padrão: {(-scores_rmse).std():.2f}')

R² por fold : [0.9595 0.9514 0.9544 0.9547 0.9557]

R²   — média: 0.9551  | desvio padrão: 0.0026
MAE  — média: 18.62 | desvio padrão: 0.29
RMSE — média: 27.91 | desvio padrão: 0.65


## 3. Comparação: validação cruzada vs teste simples

| Métrica | Teste simples (hold-out) | Cross-val média (5 folds) | Desvio padrão |
|---|---|---|---|
| R² | 0.9544 | 0.9551 | 0.0026 |
| MAE | 18.70 | 18.62 | 0.29 |
| RMSE | 28.01 | 27.91 | 0.65 |

### O resultado é consistente com o teste simples?

**Sim, muito consistente.** As diferenças entre os dois métodos são mínimas — menos de 0.1 no R² e menos de 1 unidade no MAE e RMSE. O desvio padrão entre os 5 folds é extremamente baixo (R² std = 0.0026), o que confirma que o modelo é **estável** e não depende de uma divisão específica dos dados.

Isso é um bom sinal: o modelo generaliza bem para dados não vistos independentemente de como a base é dividida.